streams/src/test/java/org/apache/kafka/streams/kstream/RepartitionTopicNamingTest.java

```diff
@@ -0,0 +1,579 @@
+    private static final String INPUT_TOPIC = "input";
+    private static final String COUNT_TOPIC = "outputTopic_0";
+    private static final String AGGREGATION_TOPIC = "outputTopic_1";
+    private static final String REDUCE_TOPIC = "outputTopic_2";
+    private static final String JOINED_TOPIC = "outputTopicForJoin";
+
+    private final String firstRepartitionTopicName = "count-stream";
+    private final String secondRepartitionTopicName = "aggregate-stream";
+    private final String thirdRepartitionTopicName = "reduced-stream";
+    private final String fourthRepartitionTopicName = "joined-stream";
+
+
+    @Test
+    public void shouldReuseFirstRepartitionTopicNameWhenOptimizing() {
+
+        final String optimizedTopology = buildTopology(StreamsConfig.OPTIMIZE).describe().toString();
+        final String unOptimizedTopology = buildTopology(StreamsConfig.NO_OPTIMIZATION).describe().toString();
+        final Pattern repartitionTopicPattern = Pattern.compile("Sink: .*-repartition");
+
+        assertThat(optimizedTopology, is(EXPECTED_OPTIMIZED_TOPOLOGY));
+        // only one repartition topic
+        assertThat(1, is(getCountOfRepartitionTopicsFound(optimizedTopology, repartitionTopicPattern)));
+        // the first named repartition topic
+        assertTrue(optimizedTopology.contains(firstRepartitionTopicName + "-repartition"));
+
+
+        assertThat(unOptimizedTopology, is(EXPECTED_UNOPTIMIZED_TOPOLOGY));
+        // now 4 repartition topic
+        assertThat(4, is(getCountOfRepartitionTopicsFound(unOptimizedTopology, repartitionTopicPattern)));
+        // all 4 named repartition topics present
+        assertTrue(unOptimizedTopology.contains(firstRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(secondRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(thirdRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(fourthRepartitionTopicName + "-left-repartition"));
+
+    }
+
+    // can't use same repartition topic name
+    @Test(expected = TopologyException.class)
+    public void shouldFailWithSameRepartitionTopicName() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder.<String, String>stream("topic").selectKey((k, v) -> k).groupByKey(Grouped.named("grouping")).count().toStream();
+        builder.<String, String>stream("topicII").selectKey((k, v) -> k).groupByKey(Grouped.named("grouping")).count().toStream();
+        builder.build();
+    }
+
+    // each KGroupedStream will result in repartition, can' reuse
+    // KGroupedStreams when specifying repartition topic names
+    // need to have separate groupByKey calls when naming repartition topics
+    // see test shouldHandleUniqueGroupedInstances below for an example
+    @Test(expected = TopologyException.class)
+    public void shouldFailWithSameRepartitionTopicNameUsingSameKGroupedStream() {
```

Comment:
Not sure why this should not be allowed? To be more precise: I understand why the code fails, however, it's very unintuitive for the user why this would fail -- looks like valid code and IMHO, users should be allowed to write this code: Both operations can reuse the same repartition topic anyway (and with optimization turned on, they will, if I don't miss anything).

Not sure if we can fix this easily to be honest, but accepting this as "by design" would not be user friendly. Maybe we can merge both repartition topics into one for this case, too?

Reply:
+1

Reply:
I agree, but IMHO it's exposing some unintuitive behavior with respect to creating multiple repartition topics.  And yes with optimizations turned on users will be able to write code in this form. 

But as it stands now, by explicitly naming a repartition I don't see how we can re-use a single `KGroupedStream` instance, as before we relied on the auto-generated names to handle the creation of multiple repartition topics.  

>Maybe we can merge both repartition topics into one for this case, too?

I'm not sure I follow, do you mean in this case we do an automatic optimization and merge repartition topics "in-line"?   If so, I'm inclined to say yes we can, but I'm thinking this may be done best in a follow-on PR.

WDYT?

Reply:
Something like this -- the idea would be to set a "flag" on the `KGroupedStream` (same for `KGroupedTable`?) after the first `count()/reduce()/aggregate()` is executed and to remember the created repartition topic. And for this case, consecutive `count()/reduce()/aggregate()` would skip creating a new changelog topic but reuse the already created one.

For backward compatibility, we would only do this if `Grouped.as` is specified. It might be a little bit hacky, but might be worth it...

Thoughts?

Reply:
I like the idea, I'll try and implement that now

Reply:
> Why limit to just when people name the repartition topic?

For backward compatibility.

For new topologies, we should not need to care, because I would assume that users turn on optimization.